src/diart/websockets.py

```diff
@@ -0,0 +1,298 @@
+
+        Parameters
+        ----------
+        client : Dict[Text, Any]
+            Client information dictionary
+        server : WebsocketServer
+            WebSocket server instance
+        """
+        client_id = client["id"]
+        logger.info(f"Client disconnected: {client_id}")
+        self.close(client_id)
+
+    def _on_message_received(
+        self, client: Dict[Text, Any], server: WebsocketServer, message: AnyStr
+    ) -> None:
+        """Process incoming client messages.
+
+        Parameters
+        ----------
+        client : Dict[Text, Any]
+            Client information dictionary
+        server : WebsocketServer
+            WebSocket server instance
+        message : AnyStr
+            Received message data
+        """
+        client_id = client["id"]
+
+        if client_id not in self._clients:
+            return
+
+        try:
+            self._clients[client_id].audio_source.process_message(message)
+        except (socket.error, ConnectionError) as e:
+            logger.warning(f"Client {client_id} disconnected: {e}")
+            self.close(client_id)
+        except Exception as e:
+            logger.error(f"Error processing message from client {client_id}: {e}")
+            # Don't close the connection for non-connection related errors
+            # This allows the client to retry sending the message
+
+    def send(self, client_id: Text, message: AnyStr) -> None:
+        """Send a message to a specific client.
+
+        Parameters
+        ----------
+        client_id : Text
+            Target client identifier
+        message : AnyStr
+            Message to send
+        """
+        if not message:
+            return
+
+        client = next((c for c in self.server.clients if c["id"] == client_id), None)
+
+        if client is None:
+            return
+
+        try:
+            self.server.send_message(client, message)
+        except (socket.error, ConnectionError) as e:
+            logger.warning(
+                f"Client {client_id} disconnected while sending message: {e}"
+            )
+            self.close(client_id)
+        except Exception as e:
+            logger.error(f"Failed to send message to client {client_id}: {e}")
+
+    def run(self) -> None:
+        """Start the WebSocket server."""
+        logger.info(f"Starting WebSocket server on {self.uri}")
+        max_retries = 3
+        retry_count = 0
+
+        while retry_count < max_retries:
+            try:
+                self.server.run_forever()
+                break  # If server exits normally, break the retry loop
+            except (socket.error, ConnectionError) as e:
+                logger.warning(f"WebSocket connection error: {e}")
+                retry_count += 1
+                if retry_count < max_retries:
+                    logger.info(
+                        f"Attempting to restart server (attempt {retry_count + 1}/{max_retries})"
+                    )
```

Comment:
We should probably `time.sleep()` for a bit before retrying

Reply:
yes, added an exponential backoff for safety